File: backend/app/models/article.py
from __future__ import annotations

import enum
from datetime import datetime
from typing import TYPE_CHECKING

from sqlalchemy import JSON, DateTime, Enum, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin

if TYPE_CHECKING:
    from app.models.pipeline_run import PipelineRun


class ArticleStatus(enum.StrEnum):
    DRAFT = "draft"
    RESEARCHING = "researching"
    OUTLINING = "outlining"
    GATE_ONE = "gate_one"
    GENERATING = "generating"
    VALIDATING = "validating"
    REVIEW = "review"
    PUBLISHED = "published"
    FAILED = "failed"


class Article(TimestampMixin, Base):
    __tablename__ = "articles"

    id: Mapped[int] = mapped_column(primary_key=True)
    slug: Mapped[str] = mapped_column(String(200), unique=True, index=True)
    title: Mapped[str] = mapped_column(String(200))
    topic: Mapped[str] = mapped_column(String(500))
    category: Mapped[str] = mapped_column(String(100), default="")
    status: Mapped[ArticleStatus] = mapped_column(
        Enum(ArticleStatus), default=ArticleStatus.DRAFT
    )
    format_id: Mapped[str] = mapped_column(String(50), default="concept")
    content_path: Mapped[str] = mapped_column(String(500), default="")
    word_count: Mapped[int] = mapped_column(default=0)
    image_count: Mapped[int] = mapped_column(default=0)
    tags: Mapped[list[str]] = mapped_column(JSON, default=list)
    reference_count: Mapped[int] = mapped_column(default=0)
    section_count: Mapped[int] = mapped_column(default=0)
    thumbnail_path: Mapped[str] = mapped_column(String(500), default="")
    published_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    published_url: Mapped[str] = mapped_column(String(500), default="")
    scheduled_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    extra_meta: Mapped[dict] = mapped_column(JSON, default=dict)

    pipeline_runs: Mapped[list[PipelineRun]] = relationship(
        back_populates="article",
        cascade="all, delete-orphan",
    )

File: frontend/src/types/article.ts
export interface Article {
  id: number;
  slug: string;
  title: string | null;
  topic: string;
  category: string | null;
  format_id: string;
  status: ArticleStatus;
  content_path: string | null;
  word_count: number;
  image_count: number;
  tags: string[];
  reference_count: number;
  section_count: number;
  thumbnail_path: string;
  created_at: string;
  updated_at: string;
}

export type ArticleStatus =
  | "draft"
  | "researching"
  | "outlining"
  | "generating"
  | "validating"
  | "review"
  | "gate_one"
  | "published"
  | "failed";

export interface ArticleCreate {
  topic: string;
  title?: string;
  format_id?: string;
}

export interface ApiResponse<T> {
  success: boolean;
  data?: T;
  error?: string;
}
